tasks/get-extension-directories-sync.js

```diff
@@ -2,21 +2,76 @@
 
 var fs = require('fs');
 var path = require('path');
-var flatMap = require('lodash/flatMap');
+var getConfig = require('./options/get-config').getConfig;
+
+function getPaths(distRelative) {
+    const from = 0;
+    const to = distRelative.lastIndexOf('/dist/');
+
+    const extensionRootPath = path.join(process.cwd(), 'node_modules', distRelative.slice(from, to));
+    const extensionSrcPath = path.join(extensionRootPath, 'src');
+    const extensionDistPath = path.join(process.cwd(), 'node_modules', distRelative);
+    const extensionCssFilePath = path.join(extensionRootPath, 'src/index.css');
+
+    return {
+        extensionRootPath,
+        extensionSrcPath,
+        extensionDistPath,
+        extensionCssFilePath,
+    };
+}
+
+function getIndexFilePath(indexFilePathRelative) {
+    const base = path.join(process.cwd(), 'dist', indexFilePathRelative);
+
+    for (const extension of ['.js', '.jsx', '.ts', '.tsx']) {
+        const filePath = base + extension;
+
+        if (fs.existsSync(filePath)) {
+            return filePath;
+        }
+    }
+
+    throw new Error('Index file not found');
+}
 
 function getExtensionDirectoriesSync() {
-    return flatMap([
-        '../scripts/extensions', // extensions from core
-        '../../../extensions', // extensions from parent repository
-    ], (relativePath) => {
-        const absolutePath = path.resolve(__dirname + '/' + relativePath);
-
-        return (fs.existsSync(absolutePath) ? fs.readdirSync(absolutePath) : [])
-            .map((extensionName) => ({extensionName, relativePath, absolutePath}))
-            .filter(
-                ({absolutePath, extensionName}) => fs.existsSync(path.join(absolutePath, extensionName, 'package.json'))
-            );
-    });
+    const config = getConfig();
+    const paths = config.importApps || config.apps || [];
+
+    // paths in `importApps` are relative client/dist
+    const indexFilePathRelative = paths.find((path) => path.startsWith('../'));
+
+    if (indexFilePathRelative == null) {
+        return [];
+    }
+
+    var indexFile = fs.readFileSync(getIndexFilePath(indexFilePathRelative), 'utf-8');
+
+    /**
+     * Regex for extracting extension id and import path(there is a capturing group for each)
+     * from index file where extensions are registered.
+     *
+     * Here's a sample of extension registration code that the regex is targeting:
+        {
+            id: 'planning-extension',
+            load: () =>
+                import('superdesk-planning/client/planning-extension/dist/planning-extension/src/extension')
+                    .then((res) => res.default),
+        }
+     */
+    var extensionRegistrationPattern = /id:\W*'(.*)',\W*load:\W*import\('(.*\/dist.+)'/g;
```

Comment:
this seems to be a bit error prone to me, what about using some json file with that config? `extension.json` or something? why do we actually need an id and not use package name?

Reply:
I agree that it's not a stable solution, but it shouldn't break often either, especially since we don't change index files often.

It's easier to setup if extensions are registered from a single place. I agree that `extension.json` would be a less error prone solution, but we would lose typescript support. It's very convenient to be able to jump to the interface and see all available extension config options. Not many extensions have their config options yet, but there will be more of them as we move more things to extensions.

The problem with using folder name as extension ID is that it needs to be passed to client-side code and it can't access the file system to retrieve the folder name.